container/apbf/filter.go

```diff
@@ -0,0 +1,624 @@
+	hash1, hash2 := siphash.Hash128(f.key0, f.key1, data)
+
+	// Attempt to find the required 'k' consecutive matches using an algorithm
+	// that reduces the average number of tests required.
+	//
+	// The algorithm works by choosing the starting position such that it leaves
+	// exactly 'k' consecutive slices to be tested, accumulates the matching
+	// sub sequences, and jumps backwards 'k' slices when a match fails.
+	//
+	// Recall that this implementation uses a ring buffer where the position
+	// index of the first slice is modified to simulate shifting, so the code
+	// below uses logical slices to map to the correct physical location within
+	// the filter data.
+	//
+	// As a further optimization, similar to when the items are added, rather
+	// than fully calculating each independent bit index for each slice via the
+	// closed formula for enhanced double hashing, this interleaves the
+	// calculation with the loop over the slices for consecutive matches which
+	// benchmarking shows is about 12% faster on average.
+	var result = false
+	f.mtx.Lock()
+	prevMatches, curMatches := uint8(0), uint8(0)
+	physicalSlice := uint16(f.l)
+	logicalSlice := (f.baseIndex + physicalSlice) % f.kPlusL
+	sliceBitOffset := uint64(logicalSlice) * f.bitsPerSlice
+	derivedIdx, acc := deriveIndex(logicalSlice, hash1, hash2)
+	for {
+		if f.isBitSet(sliceBitOffset + derivedIdx%f.bitsPerSlice) {
+			// Successful query when the required number of consecutive matches
+			// is achieved.
+			curMatches++
+			if prevMatches+curMatches == f.k {
+				result = true
+				break
+			}
+
+			// Move to the next logical slice while wrapping around the ring
+			// buffer if needed.
+			physicalSlice++
+			logicalSlice++
+			if logicalSlice == f.kPlusL {
+				logicalSlice = 0
+				sliceBitOffset = 0
+
+				// Reset the derived bit index using enhanced double hashing
+				// accordingly.
+				derivedIdx, acc = hash1, hash2
+				continue
+			}
+			sliceBitOffset += f.bitsPerSlice
+
+			// Derive the next bit index using enhanced double hashing.
+			derivedIdx += acc
+			acc += uint64(logicalSlice)
+			continue
+		}
+
+		// Nothing more to do when there are not enough slices left to achieve
+		// the required number of consecutive matches.
+		if uint16(f.k) > physicalSlice {
+			break
+		}
+
+		// Skip back the required number of matches while accumulating any
+		// matching sub sequence.
+		physicalSlice -= uint16(f.k)
+		prevMatches = curMatches
+		curMatches = 0
+
+		// Reset logical slice and derive the associated bit index using
+		// enhanced double hashing.
+		logicalSlice = (f.baseIndex + physicalSlice) % f.kPlusL
+		sliceBitOffset = uint64(logicalSlice) * f.bitsPerSlice
+		derivedIdx, acc = deriveIndex(logicalSlice, hash1, hash2)
+	}
+	f.mtx.Unlock()
+
+	return result
+}
+
+// Reset clears the filter and changes the key used in the internal hashing
+// logic to ensure a unique set of false positives as compared to prior to the
```

Comment:
s / ... compared to prior to ... / ... compared prior to ...

Reply:
I don't think it's incorrectly worded as it is, but I reworded it to the following so there is no question:

>... set of false positives versus those prior to the reset.